### ISSUE TITLE
Check Results from Llama Experiments

### ISSUE THREAD

**Issue by kerschke** [berndbischl/coseal-algsel-benchmark-repo#54]
I just added some llama experiments. However, so far I did not run the experiments with 5 replications as it is stated on our pages, but only once.
It would be cool, if you could check the results and comment on possible changes BEFORE I run the experiments again.

**Comment by berndbischl** [comment]
1 replication: yes of course lets save time now.
How long did it take? Where do you run them?

**Comment by berndbischl** [comment]
Should we also display them in a "sorted fashion"?
Hmm, probably annoying to program...

**Comment by berndbischl** [comment]
Maybe at least mark the 3 best results from the models?

**Comment by berndbischl** [comment]
Checking: Well it is a bunch of numbers :)
And in most cases we seem to be better than the single best?

What about errors? Do zero occur?

**Comment by kerschke** [comment]
Sorting them shouldn't be a problem. Just let me know, in which way you want to sort them. By "par" or by "par" within each group (baseline, classif, regr, cluster).

Running the experiments took about 15-16 hours, but I ran it on my mac - so not many cpus.
I know this wasn't very efficient, but I had to fix some minor problems on the code so the package would still pass make check and so on.

Errors didn't occur, but I tried to avoid them by removing constant features and empty classes within the training sets.

**Comment by berndbischl** [comment]
> Sorting them shouldn't be a problem. Just let me know, in which way you want to sort them. By
> "par" or by "par" within each group (baseline, classif, regr, cluster).

Thats the thing. You would like to interactively change that. Which we are not going to program now. Can you simply try to bold / color the 3 best MODEL results?

> Running the experiments took about 15-16 hours, but I ran it on my mac - so not many cpus.
> I know this wasn't very efficient, but I had to fix some minor problems on the code so the 
> package > would still pass make check and so on.

That means on a cluster - even with 5 CV reps - it takes no time at all, so we do not need to plan ahead that much. Which is good. Why dont you compute at the Dortmund SLURM site next time?

> Errors didn't occur, but I tried to avoid them by removing constant features and empty classes
> within the training sets.

Thx for doing this! I will review all code regarding the experiments today an pull all these helper functions into mlr so we do not need them anymore here.

**Comment by kerschke** [comment]
I'll try to find a way to do the highlighting.

Like I mentioned before, I had to fix some part of the code before it passed "make check". That's why I didn't add it to a cluster - even not SLURM.

Didn't you or one of your students already start on creating those helper functions for mlr?!

**Comment by berndbischl** [comment]
> Like I mentioned before, I had to fix some part of the code before it passed "make check". 
> That's why I didn't add it to a cluster - even not SLURM.

I am not sure what you mean here?  
Is there still a problem? What exactly? And if not why didnt you run on SLURM after fixing?

> Didn't you or one of your students already start on creating those helper functions for mlr?!

Yes, I need to check what you did exactly, but I think we basically cover that now. I will post details later. It cant be that mlr struggles with this, these are trivial things.

**Comment by kerschke** [comment]
Hopefully, there are no problems any more. And I didn't run on SLURM, because at the time, I fixed those issues, I already had the results :-)
Also, I would probably have to checkout and update all your packages (BBmisc, parallelMap, ParamHelpers and mlr) before I can do anything on SLURM ;-)

**Comment by berndbischl** [comment]
> And I didn't run on SLURM, because at the time, I fixed those issues, I already had the results :-)

I see.

**Comment by mlindauer** [comment]
Why is there a NA at "The cost for using all features is NA" for ASP?
And could you please add an explanation for "best", "mcp" and "par".
(and I would call it "par10" instead of "par" because the penalty factor is important)
Another question: Is "classif" a multiclass classification or something like the pair wise classification as used in satzilla? 

For the paper, I would suggest to use only approaches which were already used in successful algorithm selectors. The table is too long to use it in the paper.

**Comment by mlindauer** [comment]
And how do you compute "par10" for bbob?
There are timeouts but we don't know the cutoff.

**Comment by kerschke** [comment]
1) Looking at the feature-costs.arff for ASP, there are NAs in each of the columns - thus, the sum of those costs will result in NA.
2) True, I should add an explanation of "best" (percentage of selecting the best algorithm), "mcp" (misclassification penalty) and "par" (par10).
3) "Classif" is a multiclass classification - not a pairwise classification.
4) For the paper, I would suggest to show either the best results or a combination of the best and some common (successful) algo selectors. Maybe you guys should suggest some selectors, that should be used within the paper as I don't know all the problems (and especially not the corresponding literature).
5) What do you mean by we don't know the cutoff? We do know the cutoff time as they only depend on the dimensions. However, looking at the llama-code, there is a note that might answer your concern:
_If not specified, the performance value of the respective algorithm that was unable to solve the problem is used. That is, it is assumed that the performance value is the timeout._

**Comment by berndbischl** [comment]
> True, I should add an explanation of "best" (percentage of selecting the best algorithm), 
> "mcp" (misclassification penalty) and "par" (par10).

See #55, assigned to Lars

**Comment by berndbischl** [comment]
> For the paper, I would suggest to use only approaches which were already used in 
> successful algorithm selectors. The table is too long to use it in the paper.

Of course we can have the full table on the webpage and a subset in the paper. 
Although there is no real "standard" here. Any of these methods are valid to solve the general machine learning problem. Lets simply select the ones that work well and the ones that are well known. Well, Pascal basically said the same above, so why am I typing this?

**Comment by berndbischl** [comment]
> Why is there a NA at "The cost for using all features is NA" for ASP?

Apparently, Pascal already checked that there are indeed NAs in the costs. But note that I did not completely finish working on this. We also need to urgently display some information about this on the server pages. Also, 1 NA value in the whole cost matrix should be enough to make the sum NA, right?

**Comment by kerschke** [comment]
I completely agree. Once you have one NA, the entire sum is NA.

**Comment by kerschke** [comment]
I just added some colors to the llama-results to make them a little more readable. Hopefully, it pleases you as well :-)

**Comment by berndbischl** [comment]
> I just added some colors to the llama-results to make them a little more readable. Hopefully, it
> pleases you as well :-)

Looks good.

**Comment by berndbischl** [comment]
> Another question: Is "classif" a multiclass classification or something like the pair wise classification > as used in satzilla?

Like Pascal already said: It is "vanilla" multi-class classification. But a few further comments:

llama actually offers the methods classifyPairs. Which we COULD include in the experiments. This uses the "one-vs-one" reduction scheme of multiclass to pairwise, binary classification. But as I understand it, is does NOT weight the observation in the binary problems, but only uses majority voting in the end, so it is NOT cost-sensitive. I find using this borderline unreasonable for these reasons

a) this methodology is already used for some of the classifiers that are more or less only defined for binary problems, e.g., for SVMs this is one of the standard, simpler approaches to enable multi-class classification. Wrapping "classifyPairs" around it gains you exactly nothing.

b)  For the other methods, often a more "natural" multiclass methodology exists.

So the only gain you would get from classiflyPairs would be to enable standard multi-class classification for a classifier that can ONLY handle binary problems in R. Not many come to mind, maybe with the exception of gbm.  

Note: I am NOT talking about the WEIGHTED pairwise version here, which is another matter! This I find of course reasonable because it is a cost-sensitive method that tries to deal with the measure of interest.

**Comment by mlindauer** [comment]
> 1

ok, could you please explain the meaning of "The cost for using all features is"?
Of course, there are NAs because there are presolved instances.

> 5

maybe I still do not understand bbob. So, there is instance depending timeout, right?
However, is par10 defined and useful for such a case?
If we have a time for instance i_1 with 10 (whatever unit) and for i_2 with 10000,
the metric is highly focused on i_2 (and similar instances with a higher timeout).
The same problem exist for every metric with an average.

@berndbischl : you promised me a short description of our experimental setup (especially the preprocessing for llama)

**Comment by kerschke** [comment]
I'm not sure, whether NA is the a good value for presolved instances. IMHO it should be 0. You might have costs for the previous steps but if the calculation of those features does not cost anything because they were presolved, the costs seem to be 0 (for that step).

Coming back to BBOB: yes there are instance depending timeouts. Actually, the main measurement relERT (relative expected runtime) already uses par10, so maybe this part needs some revision.
To clarify things: the ERT might sometimes be NA (if the instance was not solved) - so let's say for a specific instance we measure the ERTs 10 (solver A), 20 (solver B) and NA (solver C). Now, the relERT standardizes all those values by the best performance, i.e. the relERTs are 1 (A), 2 (B) and 20 (C). The 20 for NA equals 10 times the worst non-NA performance of that instance.
Thus, there shouldn't be any metric issues.
However, as I mentioned at the beginning of that explanation, it might be misleading if we calculate par10 of the relERT as the relERT does not have any NAs.

Coming to your last point: actually, it's not Bernd who's not on time - it's me. I was kept busy working on the llama results and didn't start yet describing the experimental setup.

**Comment by mlindauer** [comment]
regarding the NA in cost:
lets say we have a sequential steps, i.e., s_1 -> s_2 -> s_3.
And in s_2 the instance was presolved,
s_3 was never run and it is unknown what it would cost to run the feature generation with s_3.
a cost of 0 for s_3 would be misleading because it can cost a lot more if we would have measured it.

BBOB:
What would happen if ERTs 10 (solver A) and NA for solver B and C?
relERT: 1 (A), 10 (B), 10(C)?
Would this be consistent with your example regarding the motivation of par10?

**Comment by kerschke** [comment]
BBOB: It's exactly as you assumed. Also, when all solvers had an NA as ERT, the relERT for all of them is 1.

Regarding your "NA in cost" example:
Why is it misleading to define the costs for s_3 to 0? If we didn't have any costs and still "solved" s_3 ("solved", because we don't even need it anymore), it seems to be reasonable (at least to me).
I know it's not perfect, but defining the costs to NA and thus crashing the entire calculation doesn't look like a better solution to me.

**Comment by mlindauer** [comment]
I discussed the BBOB issue yesterday with Bernd.
I'm still not convinced that it is done in a correct way. 
At least, it is not the same par10 definition as used in the other sets.
So it should be described in detail in the paper what we did.

regarding the feature costs:
also discussed with Bernd yesterday.
Our solution: If we don't use a feature step (because presolved in previous steps, crashed or whatever), then we can assume that the costs are 0 for this step. 
I do not want to replace the NAs with 0 because it is simply not the truth.
